Atomic/atomic.py

```diff
@@ -696,6 +820,407 @@ def install(self):
+                        with open(repositories_file, 'r') as rfile:
+                            repositories = rfile.read()
+                        regloc, image, tag = self._parse_imagename(list(json.loads(repositories).keys())[0])
+                        imagebranch = "%s%s-%s" % (OSTREE_OCIIMAGE_PREFIX, image, tag)
+                        input_layers = []
+                        for name in os.listdir(temp_dir):
+                            if name == "repositories":
+                                continue
+                            input_layers.append(name + "/layer.tar")
+                        self._pull_dockertar_layers(repo, imagebranch, temp_dir, input_layers)
+            finally:
+                shutil.rmtree(temp_dir)
+
+    def _check_system_ostree_image(self, repo, img, upgrade):
+        imagebranch = img.replace("ostree:", "")
+        current_rev = repo.resolve_rev(imagebranch, True)
+        if not upgrade and current_rev[1]:
+            return False
+        remote, branch = imagebranch.split(":")
+        return repo.pull(remote, [branch], 0, None)
+
+    def _check_system_oci_image(self, repo, img, upgrade):
+        regloc, image, tag = self._parse_imagename(img.replace("oci:", ""))
+        imagebranch = "%s%s-%s" % (OSTREE_OCIIMAGE_PREFIX, image.replace("sha256:", ""), tag)
+        current_rev = repo.resolve_rev(imagebranch, True)
+        if not upgrade and current_rev[1]:
+            return False
+
+        manifest = self._skopeo_get_manifest(img)
+        layers = self._get_layers_from_manifest(manifest)
+        missing_layers = []
+        for i in layers:
+            layer = i.replace("sha256:", "")
+            if not repo.resolve_rev("%s%s" % (OSTREE_OCIIMAGE_PREFIX, layer), True)[1]:
+                missing_layers.append(layer)
+                self.writeOut("Missing layer %s" % layer)
+
+        if len(missing_layers) == 0:
+            return True
+
+        layers_dir = self._skopeo_get_layers(img, missing_layers)
+        try:
+            layers = {}
+            for root, _, files in os.walk(layers_dir):
+                for f in files:
+                    if f.endswith(".tar"):
+                        layer_file = os.path.join(root, f)
+                        layer = f.replace(".tar", "")
+                        if layer in missing_layers:
+                            layers[layer] = layer_file
+
+            if (len(layers)):
+                self._import_layers_into_ostree(repo, imagebranch, manifest, layers)
+        finally:
+            shutil.rmtree(layers_dir)
+        return True
+
+    def _get_commit_metadata (self, repo, rev, key):
+        commit = repo.load_commit(rev)[1]
+        metadata = commit.get_child_value(0)
+        if key not in metadata.keys():
+            return None
+        return metadata[key]
+
+    def _checkout_system_container(self, repo, name, img, deployment, upgrade, values={}):
```

Comment:
We should probably avoid a mutable default argument here `values={}`
See http://docs.python-guide.org/en/latest/writing/gotchas/